typing/typecore.ml

```diff
@@ -4703,18 +4717,33 @@ and type_cases ?in_function env ty_arg ty_res partial_flag loc caselist =
         in
         (* Ensure that no ambivalent pattern type escapes its branch *)
         check_scope_escape pat.pat_loc env outer_level ty_arg;
-        (pat, ty_arg, (ext_env, unpacks)))
+        { typed_pat = pat;
+          pat_type_for_unif = ty_arg;
+          untyped_case = case;
+          branch_env = ext_env;
+          pat_vars = pvs;
+          unpacks;
+          contains_gadt = contains_gadt pat; }
+        )
       caselist in
+  let patl = List.map (fun { typed_pat; _ } -> typed_pat) half_typed_cases in
+  let does_contain_gadt =
+    List.exists (fun { contains_gadt; _ } -> contains_gadt) half_typed_cases
+  in
+  let ty_res, duplicated_ident_types =
+    if does_contain_gadt && not !Clflags.principal then
+      correct_levels ty_res, duplicate_ident_types half_typed_cases env
+    else ty_res, duplicate_ident_types [] env
```

Comment:
What is the meaning of this call to `duplicate_ident_types []` ?

Reply:
It is a well-typed noop: notice that `duplicate_ident_types` returns an `Env.copy_of_types`, which is abstract.